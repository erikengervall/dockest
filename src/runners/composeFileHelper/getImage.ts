import ConfigurationError from '../../errors/ConfigurationError'

const getImage = ({
  build,
  image,
  service,
}: {
  build?: string
  image?: string
  service: string
}): { image: string } | {} => {
  /**
   * If user provided an image via interface
   */
  if (typeof image === 'string' && image.length > 0) {
    return {
      image,
    }
  }

  /**
   * If user provided a build path to a Dockerfile
   */
  if (build) {
    return {}
  }

  throw new ConfigurationError(`${service} Could not determine "image" property for service`)
}

export default getImage
