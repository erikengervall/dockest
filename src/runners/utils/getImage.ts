import fs from 'fs'
import yaml from 'js-yaml'
import ConfigurationError from '../../errors/ConfigurationError'

const getImage = ({
  composeFileName,
  image: runnerConfigProvidedImage,
  mockProcess,
  service,
}: {
  composeFileName: string
  image?: string | null
  mockProcess?: any
  service: string
}): { image: string } | {} => {
  const nodeProcess = mockProcess || process

  if (runnerConfigProvidedImage === null) {
    return {}
  }

  if (typeof runnerConfigProvidedImage === 'string' && runnerConfigProvidedImage.length > 0) {
    return {
      image: runnerConfigProvidedImage,
    }
  }

  let imageFromComposeFile = null
  let dockerCompose = null
  const path = `${nodeProcess.cwd()}/${composeFileName}`

  try {
    dockerCompose = yaml.safeLoad(fs.readFileSync(path, 'utf8'))
  } catch (error) {
    throw new ConfigurationError(`Failed to parse '${path}' (${error.message})`)
  }

  try {
    imageFromComposeFile = dockerCompose.services[service].image
  } catch (error) {
    throw new ConfigurationError(`Image not found for service ${service}`)
  }

  if (typeof imageFromComposeFile === 'string' && imageFromComposeFile.length > 0) {
    return {
      image: imageFromComposeFile,
    }
  }

  throw new ConfigurationError(`${service} Invalid image found: ${imageFromComposeFile}`)
}

export default getImage
