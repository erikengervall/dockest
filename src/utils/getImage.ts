import fs from 'fs'
import yaml from 'js-yaml'
import { ConfigurationError } from '../errors'

const getImage = ({
  composeFileName,
  image: configProvidedImage,
  service,
}: {
  composeFileName: string
  image?: string
  service: string
}): string => {
  if (typeof configProvidedImage === 'string' && configProvidedImage.length > 0) {
    return configProvidedImage
  }

  let imageFromComposeFile = null
  let dockerCompose = null

  try {
    dockerCompose = yaml.safeLoad(fs.readFileSync(`${process.cwd()}/${composeFileName}`, 'utf8'))
    imageFromComposeFile = dockerCompose.services[service].image
  } catch (e) {
    throw new Error(`Failed to parse ${composeFileName}`)
  }

  if (typeof imageFromComposeFile === 'string' && imageFromComposeFile.length > 0) {
    return imageFromComposeFile
  }

  throw new ConfigurationError(`${service} Invalid image found: ${imageFromComposeFile}`)
}

export default getImage
