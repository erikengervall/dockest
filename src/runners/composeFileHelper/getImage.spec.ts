import getImage from './getImage'

describe('getImage', () => {
  const service = 'getImageService'

  describe('with user provided image', () => {
    it('should work with valid image', () => {
      const image = 'postgres:9.6'

      const result = getImage({ image, service })

      expect(result).toMatchSnapshot()
    })
  })

  describe('with user provided build', () => {
    it('should work with valid build', () => {
      const build = './path'

      const result = getImage({ build, service })

      expect(result).toMatchSnapshot()
    })
  })
})
