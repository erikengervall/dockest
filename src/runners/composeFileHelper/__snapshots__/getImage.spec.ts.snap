// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getImage with user provided build should work with valid build 1`] = `Object {}`;

exports[`getImage with user provided image should work with valid image 1`] = `
Object {
  "image": "postgres:9.6",
}
`;
